// dts-bundle-generator/dist/helpers/typescript.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getModifiers = exports.recreateRootLevelNodeWithModifiers = exports.modifiersMapToArray = exports.modifiersToMap = exports.getExportsForStatement = exports.resolveIdentifier = exports.getExportsForSourceFile = exports.getDeclarationsForSymbol = exports.isNamespaceStatement = exports.isDeclareGlobalStatement = exports.isDeclareModule = exports.isAmbientModule = exports.splitTransientSymbol = exports.getDeclarationNameSymbol = exports.getActualSymbol = exports.getNodeName = exports.hasNodeModifier = exports.isNodeNamedDeclaration = void 0;
const ts = require("typescript");
const namedDeclarationKinds = [
    ts.SyntaxKind.InterfaceDeclaration,
    ts.SyntaxKind.ClassDeclaration,
    ts.SyntaxKind.EnumDeclaration,
    ts.SyntaxKind.TypeAliasDeclaration,
    ts.SyntaxKind.ModuleDeclaration,
    ts.SyntaxKind.FunctionDeclaration,
    ts.SyntaxKind.VariableDeclaration,
    ts.SyntaxKind.PropertySignature,
];
function isNodeNamedDeclaration(node) {
    return namedDeclarationKinds.indexOf(node.kind) !== -1;
}
exports.isNodeNamedDeclaration = isNodeNamedDeclaration;
function hasNodeModifier(node, modifier) {
    const modifiers = getModifiers(node);
    return Boolean(modifiers && modifiers.some((nodeModifier) => nodeModifier.kind === modifier));
}
exports.hasNodeModifier = hasNodeModifier;
function getNodeName(node) {
    const nodeName = node.name;
    if (nodeName === undefined) {
        const modifiers = getModifiers(node);
        const defaultModifier = modifiers?.find((mod) => mod.kind === ts.SyntaxKind.DefaultKeyword);
        if (defaultModifier !== undefined) {
            return defaultModifier;
        }
    }
    return nodeName;
}
exports.getNodeName = getNodeName;
function getActualSymbol(symbol, typeChecker) {
    if (symbol.flags & ts.SymbolFlags.Alias) {
        symbol = typeChecker.getAliasedSymbol(symbol);
    }
    return symbol;
}
exports.getActualSymbol = getActualSymbol;
function getDeclarationNameSymbol(name, typeChecker) {
    const symbol = typeChecker.getSymbolAtLocation(name);
    if (symbol === undefined) {
        return null;
    }
    return getActualSymbol(symbol, typeChecker);
}
exports.getDeclarationNameSymbol = getDeclarationNameSymbol;
function splitTransientSymbol(symbol, typeChecker) {
    // actually I think we even don't need to operate/use "Transient" symbols anywhere
    // it's kind of aliased symbol, but just merged
    // but it's hard to refractor everything to use array of symbols instead of just symbol
    // so let's fix it for some places
    if ((symbol.flags & ts.SymbolFlags.Transient) === 0) {
        return [symbol];
    }
    // "Transient" symbol is kinda "merged" symbol
    // I don't really know is this way to "split" is correct
    // but it seems that it works for now ¯\_(ツ)_/¯
    const declarations = getDeclarationsForSymbol(symbol);
    const result = [];
    for (const declaration of declarations) {
        if (!isNodeNamedDeclaration(declaration) || declaration.name === undefined) {
            continue;
        }
        const sym = typeChecker.getSymbolAtLocation(declaration.name);
        if (sym === undefined) {
            continue;
        }
        result.push(getActualSymbol(sym, typeChecker));
    }
    return result;
}
exports.splitTransientSymbol = splitTransientSymbol;
/**
 * @see https://github.com/Microsoft/TypeScript/blob/f7c4fefeb62416c311077a699cc15beb211c25c9/src/compiler/utilities.ts#L626-L628
 */
function isGlobalScopeAugmentation(module) {
    return Boolean(module.flags & ts.NodeFlags.GlobalAugmentation);
}
/**
 * Returns whether node is ambient module declaration (declare module "name" or declare global)
 * @see https://github.com/Microsoft/TypeScript/blob/f7c4fefeb62416c311077a699cc15beb211c25c9/src/compiler/utilities.ts#L588-L590
 */
function isAmbientModule(node) {
    return ts.isModuleDeclaration(node) && (node.name.kind === ts.SyntaxKind.StringLiteral || isGlobalScopeAugmentation(node));
}
exports.isAmbientModule = isAmbientModule;
/**
 * Returns whether node is `declare module` ModuleDeclaration (not `declare global` or `namespace`)
 */
function isDeclareModule(node) {
    // `declare module ""`, `declare global` and `namespace {}` are ModuleDeclaration
    // but here we need to check only `declare module` statements
    return ts.isModuleDeclaration(node) && !(node.flags & ts.NodeFlags.Namespace) && !isGlobalScopeAugmentation(node);
}
exports.isDeclareModule = isDeclareModule;
/**
 * Returns whether statement is `declare global` ModuleDeclaration
 */
function isDeclareGlobalStatement(statement) {
    return ts.isModuleDeclaration(statement) && isGlobalScopeAugmentation(statement);
}
exports.isDeclareGlobalStatement = isDeclareGlobalStatement;
/**
 * Returns whether node is `namespace` ModuleDeclaration
 */
function isNamespaceStatement(node) {
    return ts.isModuleDeclaration(node) && Boolean(node.flags & ts.NodeFlags.Namespace);
}
exports.isNamespaceStatement = isNamespaceStatement;
function getDeclarationsForSymbol(symbol) {
    const result = [];
    if (symbol.declarations !== undefined) {
        result.push(...symbol.declarations);
    }
    if (symbol.valueDeclaration !== undefined) {
        // push valueDeclaration might be already in declarations array
        // so let's check first to avoid duplication nodes
        if (!result.includes(symbol.valueDeclaration)) {
            result.push(symbol.valueDeclaration);
        }
    }
    return result;
}
exports.getDeclarationsForSymbol = getDeclarationsForSymbol;
function getExportsForSourceFile(typeChecker, sourceFileSymbol) {
    if (sourceFileSymbol.exports !== undefined) {
        const commonJsExport = sourceFileSymbol.exports.get(ts.InternalSymbolName.ExportEquals);
        if (commonJsExport !== undefined) {
            const symbol = getActualSymbol(commonJsExport, typeChecker);
            return [
                {
                    symbol,
                    type: 0 /* ExportType.CommonJS */,
                    exportedName: '',
                    originalName: symbol.escapedName,
                },
            ];
        }
    }
    const result = typeChecker
        .getExportsOfModule(sourceFileSymbol)
        .map((symbol) => ({ symbol, exportedName: symbol.escapedName, type: 1 /* ExportType.ES6Named */, originalName: '' }));
    if (sourceFileSymbol.exports !== undefined) {
        const defaultExportSymbol = sourceFileSymbol.exports.get(ts.InternalSymbolName.Default);
        if (defaultExportSymbol !== undefined) {
            const defaultExport = result.find((exp) => exp.symbol === defaultExportSymbol);
            if (defaultExport !== undefined) {
                defaultExport.type = 2 /* ExportType.ES6Default */;
            }
            else {
                // it seems that default export is always returned by getExportsOfModule
                // but let's add it to be sure add if there is no such export
                result.push({
                    symbol: defaultExportSymbol,
                    type: 2 /* ExportType.ES6Default */,
                    exportedName: 'default',
                    originalName: '',
                });
            }
        }
    }
    result.forEach((exp) => {
        exp.symbol = getActualSymbol(exp.symbol, typeChecker);
        const resolvedIdentifier = resolveIdentifierBySymbol(exp.symbol);
        exp.originalName = resolvedIdentifier?.name !== undefined ? resolvedIdentifier.name.getText() : exp.symbol.escapedName;
    });
    return result;
}
exports.getExportsForSourceFile = getExportsForSourceFile;
function resolveIdentifier(typeChecker, identifier) {
    const symbol = getDeclarationNameSymbol(identifier, typeChecker);
    if (symbol === null) {
        return undefined;
    }
    return resolveIdentifierBySymbol(symbol);
}
exports.resolveIdentifier = resolveIdentifier;
function resolveIdentifierBySymbol(identifierSymbol) {
    const declarations = getDeclarationsForSymbol(identifierSymbol);
    if (declarations.length === 0) {
        return undefined;
    }
    const decl = declarations[0];
    if (!isNodeNamedDeclaration(decl)) {
        return undefined;
    }
    return decl;
}
function getExportsForStatement(exportedSymbols, typeChecker, statement) {
    if (ts.isVariableStatement(statement)) {
        if (statement.declarationList.declarations.length === 0) {
            return [];
        }
        const firstDeclarationExports = getExportsForName(exportedSymbols, typeChecker, statement.declarationList.declarations[0].name);
        const allDeclarationsHaveSameExportType = statement.declarationList.declarations.every((variableDecl) => {
            // all declaration should have the same export type
            // TODO: for now it's not supported to have different type of exports
            return getExportsForName(exportedSymbols, typeChecker, variableDecl.name)[0]?.type === firstDeclarationExports[0]?.type;
        });
        if (!allDeclarationsHaveSameExportType) {
            // log warn?
            return [];
        }
        return firstDeclarationExports;
    }
    const nodeName = getNodeName(statement);
    if (nodeName === undefined) {
        return [];
    }
    return getExportsForName(exportedSymbols, typeChecker, nodeName);
}
exports.getExportsForStatement = getExportsForStatement;
function getExportsForName(exportedSymbols, typeChecker, name) {
    if (ts.isArrayBindingPattern(name) || ts.isObjectBindingPattern(name)) {
        // TODO: binding patterns in variable declarations are not supported for now
        // see https://github.com/microsoft/TypeScript/issues/30598 also
        return [];
    }
    const declarationSymbol = typeChecker.getSymbolAtLocation(name);
    return exportedSymbols.filter((rootExport) => rootExport.symbol === declarationSymbol);
}
const modifiersPriority = {
    [ts.SyntaxKind.ExportKeyword]: 4,
    [ts.SyntaxKind.DefaultKeyword]: 3,
    [ts.SyntaxKind.DeclareKeyword]: 2,
    [ts.SyntaxKind.AsyncKeyword]: 1,
    [ts.SyntaxKind.ConstKeyword]: 1,
};
function modifiersToMap(modifiers) {
    modifiers = modifiers || [];
    return modifiers.reduce((result, modifier) => {
        result[modifier.kind] = true;
        return result;
    }, 
    // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
    {});
}
exports.modifiersToMap = modifiersToMap;
function modifiersMapToArray(modifiersMap) {
    return Object.entries(modifiersMap)
        .filter(([kind, include]) => include)
        .map(([kind]) => {
        // we don't care about decorators here as it is not possible to have them in declaration files
        return ts.factory.createModifier(Number(kind));
    })
        .sort((a, b) => {
        // note `|| 0` is here as a fallback in the case if the compiler adds a new modifier
        // but the tool isn't updated yet
        const aValue = modifiersPriority[a.kind] || 0;
        const bValue = modifiersPriority[b.kind] || 0;
        return bValue - aValue;
    });
}
exports.modifiersMapToArray = modifiersMapToArray;
function recreateRootLevelNodeWithModifiers(node, modifiersMap, newName, keepComments = true) {
    const newNode = recreateRootLevelNodeWithModifiersImpl(node, modifiersMap, newName);
    if (keepComments) {
        ts.setCommentRange(newNode, ts.getCommentRange(node));
    }
    return newNode;
}
exports.recreateRootLevelNodeWithModifiers = recreateRootLevelNodeWithModifiers;
// eslint-disable-next-line complexity
function recreateRootLevelNodeWithModifiersImpl(node, modifiersMap, newName) {
    const modifiers = modifiersMapToArray(modifiersMap);
    if (ts.isArrowFunction(node)) {
        return ts.factory.createArrowFunction(modifiers, node.typeParameters, node.parameters, node.type, node.equalsGreaterThanToken, node.body);
    }
    if (ts.isClassDeclaration(node)) {
        return ts.factory.createClassDeclaration(modifiers, newName || node.name, node.typeParameters, node.heritageClauses, node.members);
    }
    if (ts.isClassExpression(node)) {
        return ts.factory.createClassExpression(modifiers, newName || node.name, node.typeParameters, node.heritageClauses, node.members);
    }
    if (ts.isEnumDeclaration(node)) {
        return ts.factory.createEnumDeclaration(modifiers, newName || node.name, node.members);
    }
    if (ts.isExportAssignment(node)) {
        return ts.factory.createExportAssignment(modifiers, node.isExportEquals, node.expression);
    }
    if (ts.isExportDeclaration(node)) {
        return ts.factory.createExportDeclaration(modifiers, node.isTypeOnly, node.exportClause, node.moduleSpecifier, node.assertClause);
    }
    if (ts.isFunctionDeclaration(node)) {
        return ts.factory.createFunctionDeclaration(modifiers, node.asteriskToken, newName || node.name, node.typeParameters, node.parameters, node.type, node.body);
    }
    if (ts.isFunctionExpression(node)) {
        return ts.factory.createFunctionExpression(modifiers, node.asteriskToken, newName || node.name, node.typeParameters, node.parameters, node.type, node.body);
    }
    if (ts.isImportDeclaration(node)) {
        return ts.factory.createImportDeclaration(modifiers, node.importClause, node.moduleSpecifier, node.assertClause);
    }
    if (ts.isImportEqualsDeclaration(node)) {
        return ts.factory.createImportEqualsDeclaration(modifiers, node.isTypeOnly, newName || node.name, node.moduleReference);
    }
    if (ts.isInterfaceDeclaration(node)) {
        return ts.factory.createInterfaceDeclaration(modifiers, newName || node.name, node.typeParameters, node.heritageClauses, node.members);
    }
    if (ts.isModuleDeclaration(node)) {
        return ts.factory.createModuleDeclaration(modifiers, node.name, node.body, node.flags);
    }
    if (ts.isTypeAliasDeclaration(node)) {
        return ts.factory.createTypeAliasDeclaration(modifiers, newName || node.name, node.typeParameters, node.type);
    }
    if (ts.isVariableStatement(node)) {
        return ts.factory.createVariableStatement(modifiers, node.declarationList);
    }
    throw new Error(`Unknown top-level node kind (with modifiers): ${ts.SyntaxKind[node.kind]}.
If you're seeing this error, please report a bug on https://github.com/timocov/dts-bundle-generator/issues`);
}
function getModifiers(node) {
    if (!ts.canHaveModifiers(node)) {
        return undefined;
    }
    return ts.getModifiers(node);
}
exports.getModifiers = getModifiers;
